### PR TITLE
Externalizing Shipping Destinations

### PR DESCRIPTION
**Pull request by ja2-derek**
Shipping destinations - the delivery charges, names, sector placements - are used by Bobby Ray and Florist.

See  #1011 and #665.

## Summary

The order locations of Bobby Ray and the florist are now called "shipping destinations".

All delivery-related references to Drassen is generalized as the "primary destination". There must be exactly one primary destination. It is the default airport, where mercs leave, where John sends his gift and where Pablo steals.

Bobby Ray delivery to Meduna is generalized as any other destinations. There can be zero to many "non-primary" destinations. Delivery to these destinations are simple and reliable.

Flowers are delivered to the Queen if the shipping destination has a delivery sector in Meduna.

## Changes

 - Added externalization JSON
 - Extracted to translation strings for destination city names to `externalization/strings/shipping-destinations-*.json` (replaces `pDeliveryLocationStrings`)
 - Replaced all delivery-related, Drassen, SEC_B13 and Meduna references

### Constant values replaced

 - `BOBBYR_SHIPPING_DEST_SECTOR_X`, `BOBBYR_SHIPPING_DEST_SECTOR_Y`, `BOBBYR_SHIPPING_DEST_SECTOR_Z`, `BOBBYR_SHIPPING_DEST_SECTOR`
 - `BOBBYR_SHIPPING_DEST_GRIDNO`
 - `BOBBYR_SHIPMENT_ARRIVED`, `BOBBYR_SHIPMENT_ARRIVED_LENGTH` (email message)
 - `MEDUNA_ITEM_DROP_OFF_SECTOR_X`, `MEDUNA_ITEM_DROP_OFF_SECTOR_Y`, `MEDUNA_ITEM_DROP_OFF_SECTOR_Z`
 - `MEDUNA_ITEM_DROP_OFF_GRIDNO`
 - `BOBBY_R_MEDUNA_SHIPMENT`, `BOBBY_R_MEDUNA_SHIPMENT_LENGTH` (email message)

### Examples of changes

- `BR_DRASSEN` => getPrimaryShippingDestination()->locationId
- `BR_MEDUNA` => `canDeliver` && `!isPrimary`
- (`BR_DRASSEN` || `BR_MEDUNA`) => (`canDeliver` == true)

### Not covered

- HandleEquipmentLeftInDrassen: equipments are left at a hard-coded gridNo `10433` at the primary shipping destination (Drassen airport)
- No `#include` organizing/sorting. This PR is quite big already

## Adding a new destination

New shipping destinations can be added. We can create a mod to demo this if we want.

 1. Update existing or add a new shipping destination (say, Grumm) in `strings/shipping-destinations-*.json`
 1. Update or add the corresponding entry in `shipping-destinations.json`. Set `canDeliver` to true and `isPrimary` to false.
 1. Choose a delivery sector and gridNo. Most towns in game has a "delivery" building. The gridNo should be a closable crate.
 1. Update `EMAIL.edt` to add an email which notifies the player of package arrival. Update the email offset and length in `shipping-destinations.json`
 
## Testing

- Visual inspection of Laptop screens
- Ordering and receiving Bobby Ray packages to Drassen and Meduna
- Sending flowers to the Queen

### CODE REVIEW

File: assets/externalized/strings/shipping-destinations-rus.json [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+	"Метавира",		//The island of Metavira was the fictional location used by JA1
+	"Майами",		//Miami, Florida, USA (SE corner of USA)
+	"Москва",		//Moscow, USSR
+	"Нью-Йорк",		//New York, New York, USA
+	"Оттава",		//Ottawa, Ontario, Canada -- where JA2 was made!
+	"Париж",		//Paris, France
+	"Триполи",		//Tripoli, Libya (eastern Mediterranean)
+	"Токио",		//Tokyo, Japan
+	"Ванкувер"		//Vancouver, British Columbia, Canada (west coast near US border)
+]
```
Comment (lynxlynxlynx) [inline assets/externalized/strings/shipping-destinations-rus.json:19]: Are these used anywhere else? I expected them to be removed from the c++ list, hence the question.

File: src/externalized/ShippingDestinationModel.cc [hunk 2]
```diff
@@ -0,0 +1,70 @@
...
+		destGridNo = obj.GetInt("deliverySectorGridNo");
+		isPrimary = obj.getOptionalBool("isPrimary");
+	}
+
+	return new ShippingDestinationModel(
+		obj.GetInt("locationId"),
+		obj.GetInt("chargeRateOverNight"), obj.GetInt("chargeRate2Days"), obj.GetInt("chargeRateStandard"),
+		obj.getOptionalInt("flowersNextDayDeliveryCost"), obj.getOptionalInt("flowersWhenItGetsThereCost"),
+		canDeliver, isPrimary,
+		destSectorId, destSectorZ, destGridNo,
```
Comment (lynxlynxlynx) [inline src/externalized/ShippingDestinationModel.cc:40]: destSectorX and Y are not important?

File: assets/externalized/strings/shipping-destinations-rus.json (hunk 1 above)
Comment (ja2-derek) [reply assets/externalized/strings/shipping-destinations-rus.json:19]: These are used to populate the selection lists in Bobby Ray or Florist laptop screens.

File: src/externalized/ShippingDestinationModel.cc (hunk 2 above)
Comment (ja2-derek) [reply src/externalized/ShippingDestinationModel.cc:40]: `destSectorId`== `SECTOR(destSectorX, destSectorY)`. So we only need to store either one of these 2  forms and derive the other..

In game this is used to locate the 2 airports.

File: assets/externalized/strings/shipping-destinations-rus.json (hunk 1 above)
Comment (lynxlynxlynx) [reply assets/externalized/strings/shipping-destinations-rus.json:19]: Missed the removal in the second commit, nevermind.

File: src/externalized/ShippingDestinationModel.cc (hunk 2 above)
Comment (lynxlynxlynx) [reply src/externalized/ShippingDestinationModel.cc:40]: Yeah, but we use the individual coordinates in several locations directly ... besides it just being normal to initialize everything properly.

**Comment by lynxlynxlynx** [comment]
Awesome level of detail though, the game even handles sending flowers to Meduna as something special.

File: src/externalized/ShippingDestinationModel.cc (hunk 2 above)
Comment (ja2-derek) [reply src/externalized/ShippingDestinationModel.cc:40]: Updated. Now we only initialize and store X and Y in the class.

**Comment by lynxlynxlynx** [comment]
Thanks!

**Pull request merged**